[FEATURE] add setting to control centroid fill point rendering on all parts or a single part of multi-features

nirvn opened the pull request:
This PR is quite simple yet deals with an edge scenario with centroid fill that has long been a recurring irritant to me and most probably others.

Up to now, the centroid fill would draw a point for each part of a multi-polygon. This can be a desired, but at times is a nightmare. Here’s a simple example: use centroid fill to provide a visual hint as to whether a given province suffers from drought or not:
![centroid_all](https://cloud.githubusercontent.com/assets/1728657/14762091/4e11ad38-099c-11e6-960b-aa6546b8f41a.png)

Because some of the polygons have many parts (representing tiny islands within a given province in this case), the centroid points are rendered many times, making it impossible to use such layer for the above-mentioned scenario.

The PR adds a check-box to prevent centroid points from being drawn on every part of a multi-polygon. Same scenario, making use of this PR’s new option:
![centroid_one](https://cloud.githubusercontent.com/assets/1728657/14762093/57882842-099c-11e6-95ec-0b35c939ff0c.png)

To be consistent across the board, I’ve used the same boolean logic and description as what’s found with labelling.

This was made possible due to @m-kuhn ‘s recent addition of the @geometry_part_num variable, and @nyalldawson ‘s variable work. Thanks to both of you.

nirvn committed:
[FEATURE] add setting to control centroid fill point rendering
on all parts or a single part of multi-features (fixes #9199)

m-kuhn commented on src/core/symbology-ng/qgsfillsymbollayerv2.cpp:
```diff
@@ -3430,6 +3432,11 @@ void QgsCentroidFillSymbolLayerV2::renderPolygon( const QPolygonF& points, QList
 {
   Q_UNUSED( rings );
 
+  if ( context.renderContext().expressionContext().hasVariable( "geometry_part_num" ) )
```
Is this extra check required?

nirvn replied:
What would variable( "..." ).toInt() return if the variable is not present? 0? If so yeah we can safely remove the check.

m-kuhn replied:
Probably, I'd try.

m-kuhn commented:
Thanks, good idea.

What do you think about renaming ´pointOnAllParts´ to `pointOnlyOnFirstPart`? I think this reduces ambiguity with e.g. one point on the centroid of the union of all parts. That would make the effect a bit clearer of changing this setting.

And it would be great if you could add a test for this :)

nirvn commented:
Regarding your proposal about renaming, it'd indeed make it clearer but it'd divert a bit from the labelling logic. If you don't mind that I can change it.

Oh and the ever-forgotten tests! ;) Will do.

m-kuhn commented:
A tooltip on the checkbox would already be good to communicate the behavior.

nirvn commented:
@m-kuhn turns out the hasVariable() check was indeed not needed. I've updated the code there.

Also, as you suggested, I've added a tooltip to the checkbox, and also made the doc comments in the .h / .sip files clearer. Let me know if that looks ok to you.

As for the test, I'll look into it tomorrow, just noticed there isn't a centroid fill layer test yet so it'll require a bit more time (for me :smile: ) to add a brand new symbol layer test structure.

Oh, also, I'm not sure why travis' qt5 build fails (is the qgis_stylev2test timeout related to this PR?)

m-kuhn commented on src/ui/symbollayer/widget_centroidfill.ui:
```diff
@@ -42,6 +42,20 @@
+     <item>
+      <widget class="QCheckBox" name="mDrawAllPartsCheckBox">
+       <property name="text">
+        <string>Draw point point on every part of multi-part features</string>
```
> Draw point point

Sure? ;)

m-kuhn commented:
Looks good, no tests failing at the moment but looking forward to your new one coming 👍

nirvn commented:
@m-kuhn , oups oups. That's fixed fixed now ;)

DelazJ commented:
Looks really pretty, @nirvn. Great!
Question, what is identified as first part of the geometries: the biggest part or the first drawn while digitizing?

nirvn commented:
@DelazJ the first drawn part while digitizing. While labelling does pick the biggest part as its part to label, we can't do the same with centroid fill without adding a calculation cost for all symbol layer types which is probably not worth it.

nirvn commented:
@DelazJ , @m-kuhn , @nyalldawson , it might be useful to implement a geometry function that'd return a int value corresponding to the biggest part of a geometry, maybe geometry_part_biggest( @geometry ) ? Then someone could use the following function for the centroid marker size to only show a point for the biggest part: if ( @geometry_part_num = geometry_part_biggest( @geometry ), 4, 0 )

nyalldawson commented:
I'm not sure on this one. I think if we're going to add this (and I'm not discounting the usefulness of such feature!) we really should do it correctly and only draw the largest part, not the first. As it stands just drawing the first part is only a partial solution, and ultimately we'll need to add a "largest part only" setting - which means we'll be left with this extra redundant setting.

nirvn commented:
@nyalldawson no objections from my side. How would you go at it without adding a loop calculating area of all parts for all symbol layer types? Or do you think it's an acceptable cost?

nirvn commented:
@nyalldawson, if we think that a @geometry_part_biggest variable would be a useful information to expose to multi-feature symbols in general, we could add a loop for multipolygons and multilines to identify the part and expose the variable. Would that be costly?

nirvn commented:
@nyalldawson , @m-kuhn , here's a working solution to put a centroid point on the biggest part (instead of the first part). The commit adds a @geometry_part_biggest which is then used by the centroid fill's renderPolygon function.

That variable can be useful in other scenarios too, so I think it might be a worthy solution.

m-kuhn commented:
Here's another possibility which is already implemented.

    order_parts( $geometry, 'area($geometry)', False )

But `@geometry_part_biggest` is also good to have.

nirvn commented:
@nyalldawson , @m-kuhn , in case the @geometry_part_biggest option is considered too costly (i.e. due to its addition of an area() / length() for all the parts of multi-part features), I thought this could work:

`void QgsCentroidFillSymbolLayerV2::renderPolygon( const QPolygonF& points, QList<QPolygonF>* rings, QgsSymbolV2RenderContext& context )
{
  Q_UNUSED( rings );

  if ( !mPointOnAllParts )
  {
    const QgsGeometry* geom = context.feature()->constGeometry();
    const QgsGeometryCollectionV2* geomCollection = dynamic_cast<const QgsGeometryCollectionV2*>( geom->geometry() );

    double area_biggest = 0;
    int part_biggest = 1;
    unsigned int part_count = context.renderContext().expressionContext().variable( "geometry_part_count" ).toUInt();
    for ( unsigned int i = 0; i < part_count; ++i )
    {
      if ( geomCollection->geometryN( i )->area() > area_biggest )
      {
        area_biggest = geomCollection->geometryN( i )->area();
        part_biggest = i + 1;
      }
    }
    if ( context.renderContext().expressionContext().variable( "geometry_part_num" ).toInt() != part_biggest ) return;
  }

  QPointF centroid = mPointOnSurface ? QgsSymbolLayerV2Utils::polygonPointOnSurface( points ) : QgsSymbolLayerV2Utils::polygonCentroid( points );
  mMarker->renderPoint( centroid, context.feature(), context.renderContext(), -1, context.selected() );
}`

But, it crashes on the const QgsGeometry* geom line. What wrong(s) am I do here?

nirvn commented:
Oups, I simply needed to check whether there actually was a feature attached to the context.

@nyalldawson would you feel better about this alternative?

void QgsCentroidFillSymbolLayerV2::renderPolygon( const QPolygonF& points, QList<QPolygonF>* rings, QgsSymbolV2RenderContext& context )
{
  Q_UNUSED( rings );

  if ( !mPointOnAllParts )
  {
    const QgsFeature* feature = context.feature();
    if ( feature )
    {
      const QgsGeometry *geom = feature->constGeometry();
      const QgsGeometryCollectionV2* geomCollection = dynamic_cast<const QgsGeometryCollectionV2*>( geom->geometry() );

      double area_biggest = 0;
      int part_biggest = 1;
      unsigned int part_count = context.renderContext().expressionContext().variable( "geometry_part_count" ).toUInt();
      for ( unsigned int i = 0; i < part_count; ++i )
      {
        if ( geomCollection->geometryN( i )->area() > area_biggest )
        {
          area_biggest = geomCollection->geometryN( i )->area();
          part_biggest = i + 1;
        }
      }
      if ( context.renderContext().expressionContext().variable( "geometry_part_num" ).toInt() != part_biggest ) return;
    }
  }

  QPointF centroid = mPointOnSurface ? QgsSymbolLayerV2Utils::polygonPointOnSurface( points ) : QgsSymbolLayerV2Utils::polygonCentroid( points );
  mMarker->renderPoint( centroid, context.feature(), context.renderContext(), -1, context.selected() );
}

m-kuhn commented:
Not sure how big the difference will be but I think there should be a performance improvement if the area calculation is only done once and the result cached. The method will be called repeatedly for the same feature for each part. So as long as you get the same feature id from `context->feature()->id()` the variable mBiggestPartIndex should be safe to reuse and initialize only once on feature id change.


And a minor hint, lines like this often cause memory leaks:

      if ( context.renderContext().expressionContext().variable( "geometry_part_num" ).toInt() != part_biggest ) return;

If somebody inserts a `symbol2 = symbol->clone()` above and a `delete symbol2;` below, it's very easy to miss the early exit here. To avoid this:

 * linebreak after `if`
 * only `return` once (or at least only at the very beginning and very end of a method)

m-kuhn commented:
You can add a message

    Fixes #9199

to the commit message to automatically reference and close the bug in the tracker.

nirvn commented:
@m-kuhn excellent ideas. Coming up with a new commit based on that. I'll also make the code skip the area calculation for multipolygons having a single part.

nirvn commented:
@m-kuhn , @nyalldawson any better?

nyalldawson commented on src/core/symbology-ng/qgsfillsymbollayerv2.cpp:
```diff
@@ -3432,9 +3436,38 @@ void QgsCentroidFillSymbolLayerV2::renderPolygon( const QPolygonF& points, QList
+          double area_biggest = 0;
+          for ( unsigned int i = 0; i < part_count; ++i )
+          {
+            if ( geomCollection->geometryN( i )->area() > area_biggest )
```
better to store the area in a double here - you are calculating it twice (lines 3458 and 3460)

nyalldawson commented on src/core/symbology-ng/qgsfillsymbollayerv2.cpp:
```diff
@@ -3432,9 +3436,38 @@ void QgsCentroidFillSymbolLayerV2::renderPolygon( const QPolygonF& points, QList
+        }
+      }
+
+      if ( context.renderContext().expressionContext().variable( QgsExpressionContext::EXPR_GEOMETRY_PART_NUM ).toInt() != mBiggestPartIndex )
```
Could we store the part number directly as an int within QgsSymbolV2RenderContext? I really don't like using the expression contexts to pass things around like this.

nirvn replied:
@nyalldawson int mGeometryPartCount and int mGeometryPartNum variables attached to QgsSymbolV2RenderContext?

nyalldawson replied:
Yes please!

m-kuhn replied:
Or directly in the symbol layer?

nyalldawson replied:
QgsSymbolV2RenderContext would be more semantically correct, wouldn't it?

m-kuhn replied:
the context is shared between all layers, so I wouldn't pollute it unnecessarily with caches for particular purposes.

nyalldawson replied:
I'm a bit confused - we're not talking about the biggest part cache, are we? I'm referring to the part number/part count, which will always be the same for all symbol layers.

m-kuhn replied:
This is already available in the expression context and not new. Is your idea to pass them twice, once in the expression context and once in the render context? Just have to make sure that this also works below geometry generators which can work cleanly with expression context (just push and pop scopes) whereas with the rendering context this cannot just be popped again.

nyalldawson replied:
That's the intention - set it once for expressions in the expression context, and using the symbol context for non-expression usage. I'm basically wanting to avoid expression contexts being used as a shortcut way to pass values around like this.

m-kuhn replied:
The advantage of using scopes is that the variable can be isolated cleanly.

nirvn commented:
@nyalldawson , @m-kuhn , how about this? 

@nyalldawson , I found the symbol struct way a bit messy, the context option seems cleaner.

m-kuhn commented on src/core/symbology-ng/qgsfillsymbollayerv2.cpp:
```diff
@@ -3419,6 +3423,8 @@ void QgsCentroidFillSymbolLayerV2::startRender( QgsSymbolV2RenderContext& contex
   mMarker->setAlpha( context.alpha() );
   mMarker->startRender( context.renderContext(), context.fields() );
+
+  mBiggestPartFeatureId = -1;
```
How about calling this `mCurrentFeatureId`?

nirvn replied:
Done.

m-kuhn commented on src/core/symbology-ng/qgsfillsymbollayerv2.cpp:
```diff
@@ -3430,6 +3436,40 @@ void QgsCentroidFillSymbolLayerV2::renderPolygon( const QPolygonF& points, QList
+        }
+      }
+
+      if ( context.mGeometryPartNum != mBiggestPartIndex )
```
Could this be moved 3 lines down in the form

    if ( mPointOnAllParts || context.mGeometryPartNum == mBiggestPartIndex )
    {
      QPointF centroid = mPointOnSurface ? QgsSymbolLayerV2Utils::polygonPointOnSurface( points ) : QgsSymbolLayerV2Utils::polygonCentroid( points );
      mMarker->renderPoint( centroid, context.feature(), context.renderContext(), -1, context.selected() );		    mMarker->renderPoint( centroid, context.feature(), context.renderContext(), -1, context.selected() );
    }

nirvn replied:
Done.

m-kuhn commented:
The current approach looks fine to me.

nyalldawson commented on src/core/symbology-ng/qgsfillsymbollayerv2.cpp:
```diff
@@ -3430,14 +3437,51 @@ void QgsCentroidFillSymbolLayerV2::renderPolygon( const QPolygonF& points, QList
+          const QgsGeometryCollectionV2* geomCollection = dynamic_cast<const QgsGeometryCollectionV2*>( geom->geometry() );
+
+          double area = 0;
+          double area_biggest = 0;
```
area_biggest - > areaBiggest

nyalldawson commented:
Looks perfect to me to - well done!

nirvn commented:
@m-kuhn , @nyalldawson ,this is green and ready to be merged, all comments addressed. Thanks for reviewing the many iterations this went through.

NathanW2 commented:
Nice work!

On Thu, Apr 28, 2016 at 11:30 AM, Mathieu Pellerin <notifications@github.com
> wrote:

> @m-kuhn <https://github.com/m-kuhn> , @nyalldawson
> <https://github.com/nyalldawson> ,this is green and ready to be merged,
> all comments addressed. Thanks for reviewing the many iterations this went
> through.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/qgis/QGIS/pull/3026#issuecomment-215281158>
>
